File: C/14_dice.c
/**********************************************************************
* Filename    : rotaryEncoder.c
* Description : Use a Rotary Encoder.
* Author      : Robot
* E-mail      : support@sunfounder.com
* website     : www.sunfounder.com
* Update      : Cavon    2016/07/01
**********************************************************************/
#include <wiringPi.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <time.h>

#define   SDI   0   //serial data input
#define   RCLK  1   //memory clock input(STCP)
#define   SRCLK 2   //shift register clock input(SHCP)

#define   TouchPin  3

unsigned char SegCode[6] = {0x06,0x5b,0x4f,0x66,0x6d,0x7d};

unsigned char flag = 0;

void init(void)
{
	pinMode(SDI, OUTPUT); //make P0 output
	pinMode(RCLK, OUTPUT); //make P1 output
	pinMode(SRCLK, OUTPUT); //make P2 output
	pinMode(TouchPin, INPUT);
	pullUpDnControl(TouchPin, PUD_UP);

	digitalWrite(SDI, 0);
	digitalWrite(RCLK, 0);
	digitalWrite(SRCLK, 0);
}

void hc595_shift(unsigned char dat)
{
	int i;

	for(i=0;i<8;i++){
		digitalWrite(SDI, 0x80 & (dat << i));
		digitalWrite(SRCLK, 1);
		delay(1);
		digitalWrite(SRCLK, 0);
	}

		digitalWrite(RCLK, 1);
		delay(1);
		digitalWrite(RCLK, 0);
}

void randomISR(void)
{
	flag = 1;
}

int main(void)
{
	int num;

	if(wiringPiSetup() == -1){ //when initialize wiring failed,print messageto screen
		printf("setup wiringPi failed !");
		return 1; 
	}

	init();

	printf("\n");
	printf("\n");
	printf("========================================\n");
	printf("|               Dice                   |\n");
	printf("|    ------------------------------    |\n");
	printf("|        SDI connect to GPIO0          |\n");
	printf("|        RCLK connect to GPIO1         |\n");
	printf("|       SRCLK connect to GPIO 2        |\n");
	printf("|     Button Pin connect to GPIO 3     |\n");
	printf("|                                      |\n");
	printf("|     Control segment with 74HC595     |\n");
	printf("|           random number 0~6          |\n");
	printf("|    Press to supend segment 2 second  |\n");
	printf("|                                      |\n");
	printf("|                            SunFounder|\n");
	printf("========================================\n");
	printf("\n");
	printf("\n");

	if(wiringPiISR(TouchPin, INT_EDGE_FALLING, &randomISR)){
		printf("Unable to setup ISR : %s\n", strerror(errno));	
		return 1;
	}

	srand(time(NULL));
	
	while(1){
		num = rand() % 6;
		hc595_shift(SegCode[num]);
		if(flag == 1){
			printf("flag = %d, ",flag);
			printf("Pressed when %d on Segment\n", (num+1));
			delay(2000);
			flag = 0;
		}
		else{	
			delay(60);
		}
		
	}

	return 0;
}

File: Python/14_dice.py
#!/usr/bin/env python3

import RPi.GPIO as GPIO
import time
import random
from sys import version_info

if version_info.major == 3:
	raw_input = input

# Set up pins
SDI   = 17
RCLK  = 18
SRCLK = 27

TouchPin = 22

# Define a segment code from 1 to 6 in Hexadecimal
SegCode = [0x06, 0x5b, 0x4f, 0x66, 0x6d, 0x7d]

# Used to record button press
flag = 0

def print_msg():
	print ("========================================")
	print ("|               Dice                   |")
	print ("|    ------------------------------    |")
	print ("|        SDI connect to GPIO17         |")
	print ("|        RCLK connect to GPIO18        |")
	print ("|       SRCLK connect to GPIO27        |")
	print ("|     Button Pin connect to GPIO22     |")
	print ("|                                      |")
	print ("|     Control segment with 74HC595     |")
	print ("|           random number 1~6          |")
	print ("|    Press to supend segment 2 second  |")
	print ("|                                      |")
	print ("|                            SunFounder|")
	print ("========================================")
	print ("Program is running...")
	print ("Please press Ctrl+C to end the program...")
	raw_input ("Press Enter to begin\n")

def setup():
	GPIO.setmode(GPIO.BCM)
	GPIO.setwarnings(False)
	GPIO.setup(SDI, GPIO.OUT, initial=GPIO.LOW)
	GPIO.setup(RCLK, GPIO.OUT, initial=GPIO.LOW)
	GPIO.setup(SRCLK, GPIO.OUT, initial=GPIO.LOW)
	GPIO.setup(TouchPin, GPIO.IN, pull_up_down = GPIO.PUD_UP)
	GPIO.add_event_detect(TouchPin, GPIO.RISING, callback = randomISR, bouncetime = 20)

# Shift the data to 74HC595
def hc595_shift(dat):
	for bit in range(0, 8):	
		GPIO.output(SDI, 0x80 & (dat << bit))
		GPIO.output(SRCLK, GPIO.HIGH)
		time.sleep(0.001)
		GPIO.output(SRCLK, GPIO.LOW)
	GPIO.output(RCLK, GPIO.HIGH)
	time.sleep(0.001)
	GPIO.output(RCLK, GPIO.LOW)

def randomISR(channel):
	global flag
	flag = 1

def destroy():
	GPIO.cleanup()

def main():
	global flag
	print_msg()
	while True:
		num = random.randint(1,6)
		hc595_shift(SegCode[num-1])
		print (num, hex(SegCode[num-1]))
		if flag == 1:
			print ("Num: ", num)
			time.sleep(2)
			flag = 0
		else:
			time.sleep(0.01)

if __name__ == '__main__':
	setup()
	try:
		main()
	except KeyboardInterrupt:
		destroy()
